File: python/examples/sample_data_usage.py
"""
Example script demonstrating how to use the sample datasets in EngramDB.

This script shows how to:
1. Create a new database instance
2. Load a sample dataset
3. Explore the nodes and connections
"""
import os
import sys
import engramdb_py

def main():
    # Create a file-based database
    db_path = "./sample_data_example.db"
    
    print(f"Creating database at {db_path}...")
    db = engramdb_py.Database.file_based(db_path)
    
    # Load the minimal sample dataset
    print("Loading minimal sample dataset...")
    try:
        node_ids = engramdb_py.load_minimal_dataset(db)
        print(f"Loaded {len(node_ids)} nodes into the database")
    except Exception as e:
        print(f"Error loading minimal dataset: {e}")
        sys.exit(1)
    
    # Display all nodes in the database
    print("\nListing all memory nodes:")
    for node_id in db.list_all():
        node = db.load(node_id)
        title = node.get_attribute("title") or "Untitled"
        category = node.get_attribute("category") or "Uncategorized"
        
        print(f"  Node {node_id}: {title} ({category})")
        
        # List its connections
        connections = db.get_connections(node_id)
        if connections:
            print("    Connections:")
            for conn in connections:
                target_node = db.load(conn["target_id"])
                target_title = target_node.get_attribute("title") or "Untitled"
                print(f"      -> {target_title} ({conn['type']}, strength: {conn['strength']})")
    
    # Perform a vector search
    print("\nPerforming vector search for 'requirements'...")
    search_results = db.search_similar([0.9, 0.1, 0.2], 3, 0.0)
    print("Search results:")
    for node_id, similarity in search_results:
        node = db.load(node_id)
        title = node.get_attribute("title") or "Untitled"
        print(f"  {title}: similarity {similarity:.2f}")
    
    # Load the full AI bugfixing dataset (optional)
    load_full = os.environ.get("LOAD_FULL_DATASET", "0").lower() in ("1", "true", "yes")
    if load_full:
        print("\nLoading full AI bugfix workflow dataset...")
        try:
            db.clear_all()
            node_ids = engramdb_py.load_sample_dataset(db)
            print(f"Loaded {len(node_ids)} nodes from the full dataset")
            
            # Display a few nodes as examples
            print("\nSample nodes from the full dataset:")
            for i, node_id in enumerate(node_ids[:3]):
                node = db.load(node_id)
                title = node.get_attribute("title") or "Untitled"
                category = node.get_attribute("category") or "Uncategorized"
                print(f"  Node {i+1}: {title} ({category})")
            
            if len(node_ids) > 3:
                print(f"  ... and {len(node_ids) - 3} more nodes")
        except Exception as e:
            print(f"Error loading full dataset: {e}")
    
    print("\nExample completed successfully!")

if __name__ == "__main__":
    main()

File: examples/sample_data_usage.rs
use engramdb::{Database, Result};
use std::path::Path;

mod sample_dataset;

fn main() -> Result<()> {
    // Create a database with single-file storage for this example
    println!("Creating database with sample dataset...");
    let file_path = Path::new("./sample_data_example.engramdb");
    let mut db = Database::single_file(file_path)?;
    
    // Load the minimal sample dataset
    println!("Loading sample dataset...");
    let node_ids = sample_dataset::load_minimal_dataset(&mut db)?;
    println!("Loaded {} nodes into the database", node_ids.len());
    
    // Display all nodes in the database
    println!("\nListing all memory nodes:");
    for id in db.list_all()? {
        let node = db.load(id)?;
        let title = node.get_attribute("title").and_then(|attr| {
            if let engramdb::core::AttributeValue::String(s) = attr {
                Some(s.as_str())
            } else {
                None
            }
        }).unwrap_or("Untitled");
        
        let category = node.get_attribute("category").and_then(|attr| {
            if let engramdb::core::AttributeValue::String(s) = attr {
                Some(s.as_str())
            } else {
                None
            }
        }).unwrap_or("Uncategorized");
        
        println!("  Node {}: {} ({})", id, title, category);
        
        // List its connections
        let connections = db.get_connections(id, None)?;
        if !connections.is_empty() {
            println!("    Connections:");
            for conn in connections {
                println!("      -> {} ({}, strength: {})", 
                    conn.target_id, conn.type_name, conn.strength);
            }
        }
    }
    
    // Perform a vector search
    println!("\nPerforming vector search for concept similar to 'requirements'...");
    let search_results = db.search_similar(&[0.9, 0.2, 0.1], 3, 0.0, None, None)?;
    println!("Search results:");
    for (id, similarity) in search_results {
        let node = db.load(id)?;
        let title = node.get_attribute("title").and_then(|attr| {
            if let engramdb::core::AttributeValue::String(s) = attr {
                Some(s.as_str())
            } else {
                None
            }
        }).unwrap_or("Untitled");
        
        println!("  {}: similarity {:.2}", title, similarity);
    }
    
    // Load the larger sample dataset (optional)
    let load_full_dataset = std::env::var("LOAD_FULL_DATASET").map(|v| v == "1" || v.to_lowercase() == "true").unwrap_or(false);
    if load_full_dataset {
        println!("\nLoading full AI bugfix workflow dataset...");
        db.clear_all()?;
        let node_ids = sample_dataset::load_sample_dataset(&mut db)?;
        println!("Loaded {} nodes from the full dataset", node_ids.len());
        
        // Print a few representative nodes
        println!("\nSample nodes from full dataset:");
        let nodes_to_show = 3;
        for id in node_ids.iter().take(nodes_to_show) {
            let node = db.load(*id)?;
            let title = node.get_attribute("title").and_then(|attr| {
                if let engramdb::core::AttributeValue::String(s) = attr {
                    Some(s.as_str())
                } else {
                    None
                }
            }).unwrap_or("Untitled");
            
            let category = node.get_attribute("category").and_then(|attr| {
                if let engramdb::core::AttributeValue::String(s) = attr {
                    Some(s.as_str())
                } else {
                    None
                }
            }).unwrap_or("Uncategorized");
            
            println!("  Node {}: {} ({})", id, title, category);
        }
        
        println!("... and {} more nodes", node_ids.len() - nodes_to_show);
    }
    
    println!("\nExample completed successfully!");
    Ok(())
}